File: R/examples/svm_classification.R
C <- 1000;
dims <- 2;
num <- 50;

require(graphics)
require(lattice)
library("sg")

newplot <- get(getOption('device'))

meshgrid <- function(a,b) {
  list(
       x=outer(b*0,a,FUN="+"),
       y=outer(b,a*0,FUN="+")
      )
}

#set.seed(17)

trySVM <- function(c, kernel="POLY REAL 50 3 1", wireframe=FALSE) {

  sg("set_features", "TRAIN", traindat)
  sg("set_labels", "TRAIN", trainlab)
  sg("send_command", paste("set_kernel", kernel))
  sg("send_command", "init_kernel TRAIN")
  sg("send_command", "new_svm LIBSVM")
  sg("send_command", paste("c", c))
  
  sg("send_command", "svm_train")
  
  x1 <- (-49:+50)/10
  x2 <- (-49:+50)/10
  testdat <- meshgrid(x1,x2)

  testdat <- t(matrix(c(testdat$x, testdat$y),10000,2))
  
  sg("set_features", "TEST", testdat)
  sg("send_command", "init_kernel TEST")
  out <- sg("svm_classify")
  
  z <- t(matrix(out, 100, 100))
  
  svs <- sg("get_svm")$SV+1
  
  if (wireframe == TRUE)
  {
#for some reason plots only when done interactively
	  wireframe(z, shade = TRUE, aspect = c(61/87, 0.4), light.source = c(10,0,10))
  }
 else
 {
    image(x1,x2,z,col=topo.colors(1000))
    contour(x1,x2,z,add=T)
    cat(length(svs), "svs:", svs, "\n")

    posSVs <- traindat[,trainlab==+1 & (1:ncol(traindat) %in% svs)]
    negSVs <- traindat[,trainlab==-1 & (1:ncol(traindat) %in% svs)]

    pos    <- traindat[,trainlab==+1 & !(1:ncol(traindat) %in% svs)]
    neg    <- traindat[,trainlab==-1 & !(1:ncol(traindat) %in% svs)]

    matplot(posSVs[1,],posSVs[2,], pch="+", col="red",add=T, cex=1.5)
    matplot(negSVs[1,],negSVs[2,], pch="-", col="red",  add=T, cex=2.0)

    matplot(pos[1,],pos[2,], pch="+", col="black",add=T, cex=1.5)
    matplot(neg[1,],neg[2,], pch="-", col="black",add=T, cex=2.0)
 }

}

traindat <- matrix(c(rnorm(dims*num)-1,rnorm(dims*num)+1),dims,2*num)
trainlab <- c(rep(-1,num),rep(1,num))

graphics.off()
trySVM(C)
newplot()
trySVM(C, kernel="LINEAR REAL 100 1.0")
newplot()
trySVM(C, kernel="GAUSSIAN REAL 40 1.0")

File: matlab/examples/svm_classification.m
num=ceil(500*rand);
dist=rand
traindat=[rand(2,num)-dist rand(2,num)+dist];
trainlab=[-ones(1,num) ones(1,num)];

sg('set_features', 'TRAIN', traindat);
sg('set_labels', 'TRAIN', trainlab);
sg('send_command', 'set_kernel GAUSSIAN REAL 100 0.5');
sg('send_command', 'init_kernel TRAIN');
%sg('send_command', 'new_svm LIGHT');
sg('send_command', 'new_svm GNPP');
sg('send_command','svm_epsilon 1e-5')
sg('send_command', 'c 2');
sg('send_command', 'svm_train');

mi=min(traindat,2)-0.2;
ma=max(traindat,2)+0.2;
[x,y]=meshgrid(linspace(mi(1),ma(1),50), linspace(mi(2),ma(2),50));
testdat=[x(:),y(:)]';
sg('set_features', 'TEST', testdat);
sg('send_command', 'init_kernel TEST');
out=sg('svm_classify');

figure(1)
clf
out=reshape(out,50,50);
pcolor(out)
shading interp
hold on
contour(out,'k-')
colorbar
